File: javascript/leetcode/1022. Sum of Root To Leaf Binary Numbers.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumRootToLeaf = function (root) {
  return preorder(root, 0);
};

function preorder(root, val) {
  if (!root) return 0;

  val = val * 2 + root.val;
  return root.left == root.right
    ? val
    : preorder(root.right, val) + preorder(root.left, val);
}

File: c++/leetcode/1022. Sum of Root To Leaf Binary Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
  vector<vector<int>> sums;
  vector<int> num;

public:
  void preorder(TreeNode *);
  int sumRootToLeaf(TreeNode *root)
  {
    int sum = 0;
    if (!root)
      return sum;

    if (!root->right && !root->left)
      sums.push_back(vector<int>{root->val});

    num.push_back(root->val);
    preorder(root->left);
    preorder(root->right);

    for (auto i : sums)
    {
      int sz = i.size();
      int total = 0;
      for (auto j : i)
      {
        total += j * pow(2, --sz);
      }
      sum += total;
    }

    return sum;
  }
};

void Solution::preorder(TreeNode *root)
{
  if (!root)
    return;

  if (!root->left && !root->right)
  {
    num.push_back(root->val);
    sums.push_back(num);
    num.pop_back();
  }

  num.push_back(root->val);
  if (root->left)
  {
    preorder(root->left);
  }

  if (root->right)
  {
    preorder(root->right);
  }
  num.pop_back();
}